### PR TITLE
Add dgcnn model

### PR DESCRIPTION
**Pull request by ShahVandit**
## Summary

- Adds `DGCNN` (Dynamic Graph Convolutional Neural Network) adapted for EEG classification
- Treats EEG electrodes as graph nodes, dynamically builds k-NN graph in feature space at each layer
- Four EdgeConv blocks with multi-scale feature concatenation, global pooling, and MLP head
- Adapted from the original point cloud DGCNN (Wang et al.) for EEG following Song et al. (2018)

## Reference

Song, T., Zheng, W., Song, P., & Cui, Z. (2018). EEG emotion recognition using dynamical 
graph convolutional neural networks. IEEE Transactions on Affective Computing, 11(3), 532–541.
https://doi.org/10.1109/TAFFC.2018.2817622

### CODE REVIEW

File: braindecode/models/dgcnn.py [hunk 1]
```diff
@@ -0,0 +1,184 @@
...
+        self.dp1 = nn.Dropout(p=self.dropout)
+        self.linear2 = nn.Linear(512, 256)
+        self.bn7 = nn.BatchNorm1d(256)
+        self.dp2 = nn.Dropout(p=self.dropout)
+        self.final_layer = nn.Linear(256, self.n_outputs)
+
+
+    def forward(self, x: torch.Tensor) -> torch.Tensor:
+        batch_size=x.size(0)
+        x = x.transpose(1, 2)  # (B, n_chans, n_times) → (B, n_times, n_chans)
```
Comment (PierreGtch) [inline braindecode/models/dgcnn.py]:
```suggestion
        x = einops.rearrange(x, "b c t -> b t c")
```
let's use einops for clarity

File: braindecode/models/dgcnn.py [hunk 2]
```diff
@@ -0,0 +1,184 @@
...
+
+        x = get_graph_feature(x, k=self.k)
+        x = self.conv1(x)
+        x1 = x.max(dim=-1, keepdim=False)[0]
```
Comment (PierreGtch) [inline braindecode/models/dgcnn.py]: let's push these three lines into a new `DGCNNBlock` module. The way, we can factorise the code. Also we can do a for loop over the four DGCNN blocks such that the model can have an arbitrary number of blocks (parameter)

File: braindecode/models/dgcnn.py [hunk 3]
```diff
@@ -0,0 +1,184 @@
...
+        Number of EEG channels (electrodes = graph nodes).
+    chs_info : list of dict
+        Information about each channel. See :class:`mne.Info`.
+    n_times : int
+        Number of time samples per window. Used as node feature dimension.
+    input_window_seconds : float
+        Length of input window in seconds.
+    sfreq : float
+        Sampling frequency of the EEG recording.
+    k : int
```
Comment (PierreGtch) [inline braindecode/models/dgcnn.py]:
```suggestion
    block_dims: tuple[int, ...]
        Embedding dimensionality in the different `DGCNNBlocks`.
        (default: (64, 64, 128, 256))
    k : int
```

File: braindecode/models/dgcnn.py [hunk 4]
```diff
@@ -0,0 +1,184 @@
...
+    idx = idx + idx_base
+
+    idx = idx.view(-1)
+ 
+    _, num_dims, _ = x.size()
+
+    x = x.transpose(2, 1).contiguous()   # (batch_size, num_points, num_dims)  -> (batch_size*num_points, num_dims) #   batch_size * num_points * k + range(0, batch_size*num_points)
+    feature = x.view(batch_size*n_chans, -1)[idx, :]
+    feature = feature.view(batch_size, n_chans, k, num_dims) 
+    x = x.view(batch_size, n_chans, 1, num_dims).repeat(1, 1, k, 1)
```
Comment (PierreGtch) [inline braindecode/models/dgcnn.py]: the code is difficult to read here. Please use einops.rearrange and einops.repeat. It should be doable in two lines :)

File: braindecode/models/dgcnn.py [hunk 5]
```diff
@@ -0,0 +1,184 @@
...
+    inner = -2 * torch.matmul(x.transpose(2, 1), x)  # (B, n_chans, n_chans)
+    xx = torch.sum(x ** 2, dim=1, keepdim=True)       # (B, 1, n_chans)
+    pairwise_distance = -xx - inner - xx.transpose(2, 1)  # (B, n_chans, n_chans)
+    idx = pairwise_distance.topk(k=k, dim=-1)[1]      # (B, n_chans, k)
+    return idx
+
+def get_graph_feature(x, k=20, idx=None):
+    batch_size = x.size(0) # B
+    n_chans = x.size(2) # n_chans
+    x = x.view(batch_size, -1, n_chans)
```
Comment (PierreGtch) [inline braindecode/models/dgcnn.py]: einops.rearrange

File: braindecode/models/dgcnn.py [hunk 6]
```diff
@@ -0,0 +1,184 @@
...
+
+import torch
+import torch.nn as nn
+import torch.nn.functional as F
+from braindecode.models.base import EEGModuleMixin
+
+
+def knn(x, k):
+    # x: (B, n_times, n_chans)
+    inner = -2 * torch.matmul(x.transpose(2, 1), x)  # (B, n_chans, n_chans)
```
Comment (PierreGtch) [inline braindecode/models/dgcnn.py]: einops.rearrange

File: braindecode/models/dgcnn.py [hunk 7]
```diff
@@ -0,0 +1,184 @@
...
+    xx = torch.sum(x ** 2, dim=1, keepdim=True)       # (B, 1, n_chans)
+    pairwise_distance = -xx - inner - xx.transpose(2, 1)  # (B, n_chans, n_chans)
```
Comment (PierreGtch) [inline braindecode/models/dgcnn.py]: einops.rearrange

File: braindecode/models/dgcnn.py [hunk 8]
```diff
@@ -0,0 +1,184 @@
...
+
+    References
+    ----------
+    .. [dgcnn] Song, T., Zheng, W., Song, P., & Cui, Z. (2018). EEG emotion
+        recognition using dynamical graph convolutional neural networks.
+        IEEE Transactions on Affective Computing, 11(3), 532-541.
+        https://doi.org/10.1109/TAFFC.2018.2817622
+    """
+
+    def __init__(self,
```
Comment (bruAristimunha) [inline braindecode/models/dgcnn.py]: No hard-coded variable and number within the code should be a parameter.

The variable names are not super informative. 

and the n_outputs should be None, because it could be resolved on-the-fly.

File: braindecode/models/dgcnn.py [hunk 9]
```diff
@@ -0,0 +1,184 @@
...
+    x = x.view(batch_size, n_chans, 1, num_dims).repeat(1, 1, k, 1)
+    
+    feature = torch.cat((feature-x, x), dim=3).permute(0, 3, 1, 2).contiguous()
+  
+    return feature
+
+class DGCNN(EEGModuleMixin, nn.Module):
+    """DGCNN for EEG classification from Song et al. (2018) [dgcnn]_.
+
+    :bdg-success:`Graph Neural Network`
```
Comment (bruAristimunha) [inline braindecode/models/dgcnn.py]: We like to put the figure to become visually more interesting for new users.

**Review by bruAristimunha — CHANGES_REQUESTED** [review]
many thanks for the contributions, small details to adjust :)

File: braindecode/models/dgcnn.py [hunk 10]
```diff
@@ -115,70 +125,50 @@ def __init__(self,
...
+        for in_d, out_d in zip(mlp_in_dims[:-1], mlp_dims):
+            mlp_layers += [
+                nn.Linear(in_d, out_d, bias=False),
+                nn.BatchNorm1d(out_d),
+                nn.Dropout(p=drop_prob),
+                activation(),
+            ]
+        self.final_layer = nn.Linear(mlp_dims[-1], self.n_outputs)
+        mlp_layers.append(self.final_layer)
+        self.classifier = nn.Sequential(*mlp_layers)
```
Comment (bruAristimunha) [inline braindecode/models/dgcnn.py]: we usually call the last layer as final_layer

File: braindecode/models/summary.csv [hunk 11]
```diff
@@ -47,3 +47,4 @@ BENDR,General,"Classification,Embedding",250,"n_chans, n_times, n_outputs",15714
 LUNA,General,"Classification,Embedding",128,"n_chans, n_times, sfreq, chs_info",7100731,"LUNA(n_chans=22, n_times=512, sfreq=128)","Convolution,Channel,Foundation Model"
 MEDFormer,General,Classification,250,"n_chans, n_outputs, n_times",5313924,"MEDFormer(n_chans=22, n_outputs=4, n_times=1000)","Foundation Model,Convolution"
 REVE,General,Classification,200,"n_outputs, n_times, n_chans",69481476,"REVE(n_times=1000, n_outputs=4, n_chans=19)","Foundation Model,Attention/Transformer"
+DGCNN,Emotion Recognition,Classification,128,"n_chans, n_outputs, n_times",,"DGCNN(n_chans=62, n_outputs=4, n_times=128)","Graph Neural Network"
```
Comment (bruAristimunha) [inline braindecode/models/summary.csv]: it is lacking the number of parameters, you can instantiate the model and just print, should give to you the number of parameters because of the eeg classifier mixin

**Comment by bruAristimunha** [comment]
some small details, can you please fill the whats new file, apply pre-commit and then, install your branch here and run the benchmark:

https://colab.research.google.com/drive/1pWVGiwJLT53jA2m7nQ_XwvlF3v0heQ_o?usp=sharing

Here is some code that is usually used to test if the model is okay. EEGNex is an good baseline method

**Comment by ShahVandit** [comment]
I ran a quick benchmark on the BNCI2014_001 motor imagery dataset (MOABB) comparing DGCNN against EEGNeX and BrainModule: https://colab.research.google.com/drive/10PSyIqr0PRNJUWDjzkrin9k93vUxhzTl

DGCNN underperforms on this benchmark, the original paper used precomputed differential entropy (DE) features on the SEED emotion dataset, not raw EEG time series.

File: braindecode/models/dgcnn.py [hunk 12]
```diff
@@ -0,0 +1,189 @@
+# Authors: Yue Wang <yuewangx@mit.edu>
+#
```
Comment (bruAristimunha) [inline braindecode/models/dgcnn.py:3]: it is missing your name @ShahVandit

**Comment by bruAristimunha** [comment]
nice revision! 

I understand, thinking about it, I think we can generate the features and feed them into the model.

 We can generate this in many ways, such as preprocessing (in braindecode or moabb) or as a dataloader transformation.

Do you have an implementation reference for the feature? I can help you :)

**Comment by PierreGtch** [comment]
All comments on my side have been addressed, thanks @ShahVandit :)

**Comment by ShahVandit** [comment]
Hi @bruAristimunha !

Here is my approach:
1. Bandpass filter the raw EEG into 5 frequency bands (delta, theta, alpha, beta, gamma)
2. Compute log-variance per band per channel -> output shape (n_chans, 5)
3. Feed this into DGCNN with n_times=5
We could implement this as a braindecode preprocessing function.

Would love your guidance on where this fits best - preprocessing or dataloader transform.

References:
Duan et al. (2013) — "Differential Entropy Feature for EEG-based Emotion Classification" (introduced DE)- https://bcmi.sjtu.edu.cn/home/blu/papers/2013/2013-7.pdf

**Comment by ShahVandit** [comment]
Hi @bruAristimunha @PierreGtch 
I think all the requested changes have been addressed. If so, can we move forward with merging?

**Comment by ShahVandit** [comment]
Sorry, I had pushed the fix to the wrong branch! I've now applied it to the correct branch and verified all DGCNN tests pass locally. Could you please re-run CI?

**Comment by bruAristimunha** [comment]
hey @ShahVandit,

I am looking at the model, and the main problem is that its performance isn't super good. I am testing some things based on your adjusted tutorial.

**Comment by bruAristimunha** [comment]
I tested many, many things today. Looks like it is very sensitive to normalization, and if we adjust the learning parameters, we can improve a little (56% was my max), but still very, very far from the 80% for eegnex.

**Comment by ShahVandit** [comment]
Thanks for testing! I'm currently running on PhysionetMI (64 channels, closer to SEED's channel count) - the KNN graph should work better with more channels imo. Will update once I have results. Do you think it's worth merging as-is and iterating later, or should we wait for better benchmark results?

**Comment by bruAristimunha** [comment]
Some values of benchmark, and I am okay to merge. We had one sleep model in the past that was so-so, which led to extensive research comparing it with very strong baselines.

**Comment by ShahVandit** [comment]
@bruAristimunha @PierreGtch 
Apart from this, are there any models, preprocessing methods you'd like to see added to the repo? I'd love to research and discuss potential contributions when I get the opportunity :)

**Comment by ShahVandit** [comment]
Hi @bruAristimunha 
I'm running benchmarks on my end but keep running into memory/GPU issues, so it's taking longer than expected. Would it be okay to merge based on the results you've already collected? We can always revisit the benchmarks and make improvements.

**Comment by bruAristimunha** [comment]
hey @ShahVandit,

Many thanks for this PR I am testing, testing, and benchmarking, but it will eventually be merged. If you want to start another PR, I'm happy to review.

**Comment by bruAristimunha** [comment]
hey @ShahVandit,

After a long, long time of revision, I discovered the problem. The initial model implemented was the DGCNN for EEG; it was the implementation for point cloud.

Even though they share the same name, they don't actually have the same architecture.

File: docs/models/categorization/gnn.rst [hunk 13]
```diff
@@ -13,9 +13,16 @@
     :height: 56px
     :class: heading-icon no-scaled-link
 
-:bdg-light:`Graph Neural Network`
+:bdg-success:`Graph Neural Network`
```
Comment (bruAristimunha) [inline docs/models/categorization/gnn.rst]:
```suggestion
:bdg-light:`Graph Neural Network`
```

File: braindecode/models/dgcnn.py [hunk 14]
```diff
@@ -0,0 +1,402 @@
...
+        D_inv_sqrt = torch.diag(d_inv_sqrt)
+
+        I = torch.eye(self.n_chans, device=A.device)
+        return I - D_inv_sqrt @ A @ D_inv_sqrt
+
+
+class DGCNN(EEGModuleMixin, nn.Module):
+    r"""DGCNN for EEG classification from Song et al. (2018) [dgcnn]_.
+
+    :bdg-success:`Graph Neural Network`
```
Comment (bruAristimunha) [inline braindecode/models/dgcnn.py]:
```suggestion
    :bdg-light:`Graph Neural Network`
```

**Review by bruAristimunha — APPROVED** [review]

**Comment by bruAristimunha** [comment]
Now it is working, after many hours of studying the paper and finding the original source code. If we want to continue on the graph neural network, it could be interesting to implement the top 5-7 more famous methods.

File: braindecode/models/dgcnn.py [hunk 15]
```diff
@@ -0,0 +1,409 @@
...
+import torch
+import torch.nn as nn
+import torch.nn.functional as F
+from sklearn.metrics import pairwise_distances
+from sklearn.neighbors import kneighbors_graph
+
+from braindecode.models.base import EEGModuleMixin
+from braindecode.models.util import extract_channel_locations_from_chs_info
+
+
```
[the remark this answers is not in the document]
Comment (bruAristimunha) [reply braindecode/models/dgcnn.py:23]: We have Skorch

File: braindecode/models/dgcnn.py [hunk 16]
```diff
@@ -0,0 +1,409 @@
...
+        self.activation = activation()
+        self.drop_prob = drop_prob
+
+        # Learnable adjacency W* (Section 3.1, Algorithm 1)
+        # Use self.chs_info (populated by EEGModuleMixin with defaults
+        # when chs_info=None) so that the adjacency can always be built.
+        self.learned_adj = _LearnableAdjacency(
+            n_chans=self.n_chans,
+            chs_info=self.chs_info,
+            n_neighbors=n_neighbors,
```
[the remark this answers is not in the document]
Comment (bruAristimunha) [reply braindecode/models/dgcnn.py:446]: Our mixin already do this

File: braindecode/models/dgcnn.py [hunk 17]
```diff
@@ -0,0 +1,409 @@
...
+class DGCNN(EEGModuleMixin, nn.Module):
+    r"""DGCNN for EEG classification from Song et al. (2018) [dgcnn]_.
+
+    :bdg-light:`Graph Neural Network` :bdg-dark-line:`Channel`
+
+    .. figure:: ../docs/_static/model/DGCNN.gif
+        :align: center
+        :alt: DGCNN Architecture
+        :width: 600px
+
```
[the remark this answers is not in the document]
Comment (bruAristimunha) [reply braindecode/models/dgcnn.py]: Good point, I didn't check

File: docs/conf.py [hunk 18]
```diff
@@ -19,8 +19,14 @@
...
 # documentation root, use os.path.abspath to make it absolute, like shown here.
 
-import inspect
 import os
+
+# Enable docstring inheritance before any braindecode imports so that
+# NumpyDocstringInheritanceInitMeta merges parent parameters into child
+# class docstrings (e.g. EEGModuleMixin params into every model).
+os.environ["DOCSTRING_INHERITANCE_ENABLE"] = "1"
+
```
Comment (bruAristimunha) [inline docs/conf.py]:
```suggestion
```
